### PR TITLE
Initial commit for serve.run

### PR DESCRIPTION
**Pull request by simjay**
## Why are these changes needed?

A `serve.run` cli tool will help development of serve by reducing re-deploy time

## Related issue number

#16094

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — COMMENTED** [review]
Thanks @simjay this is amazing. Maybe we can try to add some tests? You can add them in `serve/tests/test_cli.py`

File: python/setup.py [hunk 1]
```diff
@@ -144,6 +144,7 @@
     "redis >= 3.5.0",
     "opencensus",
     "prometheus_client >= 0.7.1",
+    "watchgod"
```
[the remark this answers is not in the document]
Comment (simjay) [reply python/setup.py:147]: Will do!

File: python/ray/serve/scripts.py [hunk 2]
```diff
@@ -56,3 +57,23 @@ def start(http_host, http_port, http_location):
...
+@click.option(
+    "--verbose",
+    default=True,
+    required=False,
+    type=bool,
+    help="Enable verbose mode.")
+def run(filepath, verbose):
+    for _ in watch(filepath):
+        if verbose:
+            print("[SERVE] Deployment script is modified. Redeploying...\n")
```
[the remark this answers is not in the document]
Comment (simjay) [reply python/ray/serve/scripts.py:78]: Will do!

File: python/ray/serve/scripts.py [hunk 3]
```diff
@@ -56,3 +57,23 @@ def start(http_host, http_port, http_location):
...
+        exec(open(filepath).read())
```
[the remark this answers is not in the document]
Comment (simjay) [reply python/ray/serve/scripts.py:79]: Will do this. Just out of curiosity, could you elaborate a bit more on `exec won't change the import so imported backend defs won't be updated.`? Still a bit new to Ray ecosystem :)

**Comment by edoakes** [comment]
Looks like `test_cli` is still failing here:

https://buildkite.com/ray-project/ray-builders-pr/builds/6825#295ae278-8a3e-4e98-aa2b-b55653edea4f/258-3974

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by simon-mo** [comment]
Thanks for the prototype! @shrekris-anyscale will pick this up in his CLI work.

**Pull request closed without merging**